### ISSUE TITLE
fairseq generate shorter translations than tensor2tensor

### ISSUE THREAD

**Issue by JiqiangLiu** [pytorch/fairseq#500]
Recently, I found that fairseq's transformer generate shorter translation than tensor2tensor. My training data is Chinese-English of CWMT2018. I run the transformer base model. And my valid is wmt17-dev , whose BP is 0.946 when length penalty is 1.5. But in the tensor2tensor when length penalty is 1.5 , BP is close to 1.0. Why is that?

**Comment by huihuifan** [comment]
Are all of the other training parameters the same?

**Comment by JiqiangLiu** [comment]
I confirm!

**Comment by myleott** [comment]
Can you please share some more details:
- the exact training and generate command you ran
- is the BLEU comparable?

I'm not familiar with how decoding is implemented in t2t, but one thing we were careful with in fairseq is that you get the same results regardless of the batch size during decoding. This means that you need to stop generation for each sentence immediately after you finalize a sufficient number of hypotheses. You can disable this behavior by adding the `--no-early-stop`. In fairseq we also compare normalized hypothesis scores (i.e., we compare average per-token logprobs instead of the sum). You can turn that off with `--unnormalized`, which will produce longer outputs.

**Comment by JiqiangLiu** [comment]
@myleott my training command is `python train.py $data_dir
--arch transformer_wmt_en_de
--optimizer adam
--adam-betas (0.9,0.98)
--clip-norm 0.0
--lr-scheduler inverse_sqrt
--warmup-init-lr 1e-07
--warmup-updates 4000
--lr 0.0007
--min-lr 1e-09
--criterion label_smoothed_cross_entropy
--label-smoothing 0.1
--log-interval 100 
--max-update $max_update
--save-dir $model_dir
--no-progress-bar
--update-freq 2
--max-tokens 4096 
--distributed-world-size $distributed_world_size
$share_decoder_input_output_embed`, and generate command is `python generate.py $data
--path $checkpoint
--batch-size 32
--beam 12 
--lenpen 1.5 
--gen-subset test
$remove_bpe 
$raw_text`.
Usually I use mult-bleu.perl to calculate. When I use --unnormalized, I found that BP=0.851 which means the shorter translation.

**Comment by myleott** [comment]
Ah yes, unnormalized should bias towards shorter outputs, my mistake.

Is the valid loss and BLEU comparable between the two models? How much variability is there due to random initialization (you can try varying the --seed flag)?

Also I noticed you don’t have dropout specified in your training command. That will use the architecture default of 0.1. Is that intended and does it match what you did in t2t?

It’s not clear whether the differences you’re observing between fairseq and t2t are due to differences in the hyperparams (e.g., dropout, or maybe batch size since I noticed you’re using update-freq), or due to randomness in model training (i.e., you’d get similar magnitude of differences by varying the random seed), or subtle differences in the decoding algorithms between fairseq and t2t (eg., maybe length penalty is implemented differently?).

We have been able to reproduce their results exactly on En-De, so I suspect it’s something related to hyperparams or inherent randomness.

**Comment by myleott** [comment]
Okay, I just found this comment in the tensor2tensor code: “Length penalty is given by = (5+len(decode)/6) ^ -\alpha. Pls refer to https://arxiv.org/abs/1609.08144.”

In fairseq we just use len(decode)^\alpha.

You’re can update fairseq to match tensor2tensor and see how that changes things: https://github.com/pytorch/fairseq/blob/184629a721adf84a4fbd1fd5db8306c49da7685d/fairseq/sequence_generator.py#L263

**Comment by JiqiangLiu** [comment]
Thank you very much. I have already reproduce en-de using the fairseq, which is little difference from tensor2tensor. But for the task of Chinese-English, I observe that different behavior, which is interesting. I will try to update fairseq to see how that changes things.

**Comment by JiqiangLiu** [comment]
Hi, is there a length penalty for each step of fairseq decoding? I don't think I see it.

**Comment by huihuifan** [comment]
For Chinese-English, we have some pretrained models available here: https://github.com/pytorch/fairseq/tree/master/examples/pay_less_attention_paper based on a paper we worked on: https://openreview.net/forum?id=SkVhlh09tX

For the length penalty at each step, we do not have this implemented, but you could add it in sequence_generator.py

**Comment by JiqiangLiu** [comment]
Thank you very much!

**Comment by GeorgeS2019** [comment]
@huihuifan  I now promote tensor2tensor Zh to En as it is not possible to make use of the [pre-trained model](https://dl.fbaipublicfiles.com/fairseq/models/dynamicconv/wmt17.zh-en.lightconv-glu.tar.bz2)
